### PR TITLE
ListenableFuture for authorize index action

### PR DESCRIPTION
**Pull request by ywangd**
Loading authorizedIndices can become more expensive with the increase of
index names (generally in proportion to the cluster size). Computing
indicesAccessControl can also become more expensive with the increase of
requested indices, e.g. a wildcard search targeting all indices in the
cluster. Worse, the computations are repeated for every request even
when they are identical. This means high concurrency of user requests,
e.g. searches, can potentially spend a lot of time in authorization
which in turn blocks transport worker and leads to cluster instability.

This PR changes both calculations to use ListenableFuture so that the
expensive work is only done once for requests (REST or Transport) that
are (1) identical in terms of index authorization and (2) trying to go
through authorization process concurrently, e.g. Field Caps requests
sent by Kibana. When N such requests are trying to perform
authorization, only the first one will actually do the work and other
requests will simply register a listener to the result. Once the first
request has completed its work, the result is shared among the waiting
requests.

The first request is also responsible for clearing out the
listenableFuture right after it completion. The advantage is that we do
not need extra cache managament. The tradeoff is that the effective TTL
is really short (essentially the elapsed time of the associated method)
and is less effective for maximizing caching effects. Overall, the
current approach is more preferrable for its simplicity and benefits
towards high concurrent load.

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java [hunk 1]
```diff
@@ -354,6 +369,48 @@ public void authorizeIndexAction(RequestInfo requestInfo, AuthorizationInfo auth
         }
     }
 
+    private final Map<IndicesAccessControlCacheKey, ListenableFuture<IndicesAccessControl>> indicesAccessControlLookup =
+        new ConcurrentHashMap<>();
+
+    private void buildIndicesAccessControlAsync(
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:375]: This change will have an interplay with #78327 if both gets merged. In that case, my plan is to conditionally apply both optimisations: the optimisation here will only be picked if it is not covered by #78327, e.g. when the request does *not* target a single index.

**Comment by ywangd** [comment]
I thought about apply the ListenableFuture pattern to `authorizeIndexAction` so it does not have to be applied separately to the two methods of `loadAuthorizedIndices` and `buildIndicesAccessControl`. But the issues are:
* There is no useful equality comparison for a `TransportRequest` which makes creating a cacheKey tricky and error prone.
* `authorizeIndexAction` has a `indicesAsyncSupplier` as its argument. There is again no useful equality comparison for it. It is in theory possible to replace it with its underlying variables, but that leads a rather big change which is not justified and unlikely to get done for this current release.

**Comment by ywangd** [comment]
Another place is apply this pattern is around `IndicesAndAliasesResolver#resolve` which can be expensive if the requested indices have wildcards, i.e. requests directly related to REST level requests. I didn't do it for this PR to keep the change size under control. If we agree on the approach in this PR, I can add it as a follow-up.

**Comment by ywangd** [comment]
@elasticmachine update branch

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java [hunk 2]
```diff
@@ -354,6 +369,48 @@ public void authorizeIndexAction(RequestInfo requestInfo, AuthorizationInfo auth
...
+            listener.map((IndicesAccessControl indicesAccessControl) -> new IndexAuthorizationResult(true, indicesAccessControl));
+
+        if (authorizationInfo instanceof RBACAuthorizationInfo) {
+            final Role role = ((RBACAuthorizationInfo) authorizationInfo).getRole();
+            final Tuple<IndicesPermission, IndicesPermission> indicesPermissionTuple = getRoleIndicesPermissions(role);
+            final Set<String> requestedIndices = Set.copyOf(resolvedIndices.getLocal());
+            final IndicesAccessControlCacheKey cacheKey = new IndicesAccessControlCacheKey(metadata.version(),
+                action, requestedIndices,
+                indicesPermissionTuple.v1(),
+                indicesPermissionTuple.v2());
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:392]: It looks to me like a lot of code to compute a cache key; it's hard to follow and counter-productive to the performance goal of this all.

I think we should work with the fact that the `Role`, `resolvedIndices.getLocal`, `action` and `metadata.getIndicesLookup()` are immutable (or effectively immutable, maybe assert or test it).
In this case, the cache key class's `equals` method would use the `==` test for the corresponding members; analogous with `System#identityHashCode`.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java [hunk 3]
```diff
@@ -354,6 +369,48 @@ public void authorizeIndexAction(RequestInfo requestInfo, AuthorizationInfo auth
...
+
+    private void buildIndicesAccessControlAsync(
+        String action,
+        AuthorizationInfo authorizationInfo,
+        ResolvedIndices resolvedIndices,
+        Metadata metadata,
+        ActionListener<IndexAuthorizationResult> listener) {
+
+        final ActionListener<IndicesAccessControl> wrappingListener =
+            listener.map((IndicesAccessControl indicesAccessControl) -> new IndexAuthorizationResult(true, indicesAccessControl));
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:383]: Nit, can we "cache" the `IndexAuthorizationResult` object instead of the indicesAccessControl?

**Review by albertzaharovits — COMMENTED** [review]
I think this looks good and promising, and we can maybe later investigate making the cache entries more long lived. 

From an implementation point of view, I think we can do better to make the cache key computation simpler, and maybe more isolated from the core logic.

From a performance point of view, I think we could set the `concurrencyLevel` parameter on the map to the number of transport worker threads, since we are positive every such thread will now do a `computeIfAbsent` on this map (it's more coupling than I usually like, but I think we have to find a heuristic for it - I don't have good suggestion here - but neither great expectations).

I'll take another look after you manage to simplify the cache key computation a bit.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java (hunk 2 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java:392]: I don't quite follow. Maybe I am missing something, but how is it possible to use `==` to compare `action` or `resolvedIndices.getLocal`? For example, even if two `action` strings have the same value, they are different objects belong to two different HTTP worker threads? 

Comparing identity of `metadata.getIndicesLookup()` may be viable, but it does not seem to be more convenient or more reliable than comparing a simple version number?

As for `Role`, the issue is that the roles for API keys are [constructed on the fly](https://github.com/elastic/elasticsearch/blob/eeb09f0248765c7b24a4ba8129e29141e1e7c956/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java#L302) for each request. So even the same API key will get two equavilent `Role` objects but with different identities. To get around it, the cacheKey here uses only the `IndicesPermission` component of a `Role` and the comparison for `IndicesPermission` is an object identity.

**Comment by ywangd** [comment]
> I think this looks good and promising, and we can maybe later investigate making the cache entries more long lived.
> 
> From an implementation point of view, I think we can do better to make the cache key computation simpler, and maybe more isolated from the core logic.

With #78327 replaced by the changes in core (#78508), together with #78321, we are in a very good position for Index/Shard level requests, I feel fairly confident to say that we don't need any more optimisation for them in 7.16. Therefore, I'd like to rework this PR so it targets more specifically to REST level requests, or more technically, requests that are `Replaceble`. It was indeed raised mainly for REST level requests, but it also had index/shard level requests in mind, which is no longer necesesary. I need sometime to think through it first (maybe will ping you for ideas). Thanks!

> From a performance point of view, I think we could set the `concurrencyLevel` parameter on the map to the number of transport worker threads, since we are positive every such thread will now do a `computeIfAbsent` on this map (it's more coupling than I usually like, but I think we have to find a heuristic for it - I don't have good suggestion here - but neither great expectations).
> 

This is a good point. I'll incorporate it in the rework.

**Comment by ywangd** [comment]
I have reworked this PR based on the recent development and feedback. It is now ready for another around. A few key points:
* It now focuses on requests that are `Replaceable`. This is a better compliment given #78321 and #78508
* It now caches the entire indexAuthorization instead of trying to caching each component (authorizedIndices, resolvedIndices, indicesAccessControl) separately. I think this is overall better for efficency.
* The listenableFuture now branches to  the `generic` thread pool instead of executing directly (which was wrong)
* It now uses actual `Cache` instead of concurrentHashMap for better cache control. The cache can also be disabled now.

**Review by albertzaharovits — COMMENTED** [review]
I acknowledge the amount of effort here, but I worry about the added complexity in deciding when to cache and what to do with a cached value.

The complexity will make it even more nightmarish to maintain, and I don't think this overall is a great performance boost, because we now introduce atomics and forking on thread pools which, for easy-to-authorize requests it actually amounts to a performance penalty.

**Review by tvernum — COMMENTED** [review]
My inclination is similar to @albertzaharovits's here. This introduces a fair bit more complexity into code that is already far too complex to safely deal with.
What sort of performance gain do we expect here? It's hard to evaluate the tradeoff without some real data.

**Comment by tvernum** [comment]
>> we now introduce atomics and forking on thread pools which, for easy-to-authorize requests it actually amounts to a performance penalty.

> If the request is easy-to-authorize, it's likely that the caching won't even trigger.

But the cache management code (like the AtomicBoolean to check whether an entry exists in the cache) does still trigger. For a hypothetical node where the caching is never useful (because it literally _never_ has concurrent requests over the same indices) this code will be slower than the old node because the cache exists & needs to be managed, but adds no value.  
Is that 1% slower? or 10% ? 0.1%?  
I suspect it's not going to be much, but given we could drop double digit % from some code just by avoiding collection resizing (which was not an obvious problem from reading the code) it's a question that we would want to have an answer for - what % slow down in "normal" request patterns is acceptable to optimize the "slow and duplicated" ones? 

> the ratio is about 3:7:12 for loadAuthorizedIndices, buildIndicesAccessControl and resolveIndicesAndAliases

This is where I think tradeoff questions come into play. If we can find a way to cache `resolveIndicesAndAliases` in a way that is less complex than the proposal we have here, then that might be a win. It could be a 50% improvement, for less code complexity.
Maybe there is no such option - but I think the conclusion from the attempting you've made to add caching here is that it's hard to do in a way that we're comfortable with and we need to explore other options.

**Comment by ywangd** [comment]
Closing this as we agreed that it is not an overall good tradeoff.
